lib/pure/endians.nim

```diff
@@ -98,16 +124,734 @@ else:
+
+  var value = cast[ptr T](buf)[]
+  cast[ptr T](buf)[] = swapEndian(value)
+
+
+func swapEndian*(T: typedesc[SomeNumber], buf: var openArray[byte],
+                 startIndex: Natural) {.inline, since: (1,1).} =
+  ## Swaps the byte order of a number stored as a byte sequence in `buf`
+  ## starting at `startIndex`. The type of the number must be passed in as an
+  ## argument.
+  runnableExamples:
+    var buf: array[6, byte]
+    buf[0] = 0xff
+    buf[1] = 0xca
+    buf[2] = 0xfe
+    buf[3] = 0x42
+    swapEndian(uint16, buf, 1)
+    doAssert buf[0] == 0xff
+    doAssert buf[1] == 0xfe
+    doAssert buf[2] == 0xca
+    doAssert buf[3] == 0x42
+
+  let i = startIndex
+  when nimvm:
+    when sizeof(T) == 1:
+      discard
+
+    elif sizeof(T) == 2:
+      swap(buf[i],   buf[i+1])
+
+    elif sizeof(T) == 4:
+      swap(buf[i],   buf[i+3])
+      swap(buf[i+1], buf[i+2])
+
+    elif sizeof(T) == 8:
+      swap(buf[i],   buf[i+7])
+      swap(buf[i+1], buf[i+6])
+      swap(buf[i+2], buf[i+5])
+      swap(buf[i+3], buf[i+4])
+  else:
+    assert i < buf.len
+    assert i + sizeof(T) <= buf.len
+    swapEndian(T, buf[startIndex].addr)
+
+
+func toBE*[T: SomeNumber](value: T): T {.inline, since: (1,1).} =
+  ## Converts a number so when it's written to a file or a stream as a byte
+  ## sequence using endianness unaware I/O, the resulting byte sequence will
+  ## be in big-endian byte order.
+  ##
+  ## Consider using `toBytesBE` instead to make the intent clearer.
+  when system.cpuEndian == bigEndian: value
+  else: swapEndian(value)
+
+func toLE*[T: SomeNumber](value: T): T {.inline, since: (1,1).} =
+  ## Converts a number so when it's written to a file or a stream as a byte
+  ## sequence using endianness unaware I/O, the resulting byte sequence will
+  ## be in little-endian byte order.
+  ##
+  ## Consider using `toBytesLE` instead to make the intent clearer.
+  when system.cpuEndian == littleEndian: value
+  else: swapEndian(value)
+
+func fromBE*[T: SomeNumber](value: T): T {.inline, since: (1,1).} =
+  ## Converts a big-endian number read from a file or a stream using
+  ## endianness unaware I/O to native byte order, so it can be used for
+  ## arithmetic calculations in a program. This is equivalent to `toBE`,
+  ## it's only included for readability.
+  ##
+  ## Consider using `fromBytesBE` instead to make the intent clearer.
+  toBE(value)
+
+func fromLE*[T: SomeNumber](value: T): T {.inline, since: (1,1).} =
+  ## Converts a little-endian number read from a file or a stream using
+  ## endiannes unaware I/O to native byte order, so it can be used for
+  ## arithmetic calculations in a program. This is equivalent to `toLE`,
+  ## it's only included for readability.
+  ##
+  ## Consider using `fromBytesLE` instead to make the intent clearer.
+  toLE(value)
+
+func toBytesBE*[T: SomeNumber](value: T, buf: pointer)
+    {.inline, since: (1,1).} =
+  ## Writes a number to a memory location as a big-endian byte sequence.
```

Comment:
``pointer`` is unnecessarily unsafe and low level.

Reply:
just found a bug related to this that could easily have been avoided with good API: https://github.com/status-im/nim-eth/blob/bc6c9819166fd2aba242aee06bd7ea59725ebcac/eth/p2p/auth.nim#L195

notice how the address of an `int` is passed in - this works on little endian platforms but not on big endian platforms, which is extra funny

Reply:
this is why we wrote `endians2` as well: the existing API is needlessly unsafe for no good reason